src/Raven.Client/Documents/Session/AsyncSessionDocumentCounters.cs

```diff
@@ -0,0 +1,147 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="AsyncSessionDocumentCounters.cs" company="Hibernating Rhinos LTD">
+//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Raven.Client.Documents.Session
+{
+    public class AsyncSessionDocumentCounters : SessionCountersBase, ICountersSessionOperationsAsync
+    {
+        public AsyncSessionDocumentCounters(InMemoryDocumentSessionOperations session, string documentId) : base(session, documentId)
+        {
+        }
+
+        public AsyncSessionDocumentCounters(InMemoryDocumentSessionOperations session, object entity) : base(session, entity)
+        {
+        }
+
+        public async Task<Dictionary<string, long>> GetAllAsync()
+        {
+            if (Session.CountersByDocId.TryGetValue(DocId, out var cache))
+            {
+                if (cache.GotAll)
+                {
+                    var result = new Dictionary<string, long>();
+                    foreach (var kvp in cache.Values)
+                    {
+                        if (kvp.Value.HasValue)
+                        {
+                            result.Add(kvp.Key, kvp.Value.Value);
+                        }
+                    }
+
+                    return result;
+                }
+
+            }
+            else
+            {
+                cache = new InMemoryDocumentSessionOperations.CountersCache
+                {
+                    Values = new Dictionary<string, long?>()
+                };
+            }
+
+            cache.GotAll = true;
+            Session.CountersByDocId[DocId] = cache;
```

Comment:
This should be add _after_ the call, because an exception may be thrown here.

Reply:
fixed